### PR TITLE
[hotfix] Add purpose-based deletion grace period for backup infrastructure

### PR DESCRIPTION
**Pull request by shreyas-s-rao**
Signed-off-by: Shreyas Rao <shreyas.sriganesh.rao@sap.com>

**What this PR does / why we need it**:
This PR adds support for setting the backup infrastructure deletion grace period per shoot purpose via controller manager config, via new field `deletionGracePeriodHoursByPurpose`. This PR also modifies the field `deletionGracePeriodDays` for backupInfrastructure controller to `deletionGracePeriodHours`, to provide more fine-grained control over backup bucket retention period.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
This feature serves as a mitigation for the bucket quota limit issues on AWS.
@swapnilgm @adracus @zanetworker 

**Release note**:

```noteworthy operator
The field `deletionGracePeriodHoursByPurpose` is now introduced for the backupInfrastructure controller to specify different deletion grace period values per shoot purpose.
```

### CODE REVIEW

File: pkg/controllermanager/controller/backupinfrastructure/backup_infrastructure_control.go [hunk 1]
```diff
@@ -415,8 +415,20 @@ func (c *defaultControl) removeFinalizer(op *operation.Operation) error {
 	})
 }
 
-func computeGracePeriod(config *config.ControllerManagerConfiguration) time.Duration {
-	return time.Hour * 24 * time.Duration(*config.Controllers.BackupInfrastructure.DeletionGracePeriodDays)
+func computeGracePeriod(config *config.ControllerManagerConfiguration, backupInfrastructure *gardenv1beta1.BackupInfrastructure) time.Duration {
+	var deletionGracePeriod time.Duration
+
+	shootPurpose := backupInfrastructure.ObjectMeta.Annotations[gardencorev1alpha1.GardenPurpose]
```
Comment (swapnilgm) [inline pkg/controllermanager/controller/backupinfrastructure/backup_infrastructure_control.go]: Simply this. 

```suggestion
      shootPurpose := backupInfrastructure.ObjectMeta.Annotations[gardencorev1alpha1.GardenPurpose]
	if deletionGracePeriodHours, ok := deletionGracePeriodHoursByPurpose[shootPurpose]; ok {
	  return  time.Hour * time.Duration(deletionGracePeriodHours)
	}
	return time.Hour * time.Duration(*config.Controllers.BackupInfrastructure.DeletionGracePeriodHours)
```

**Review by zanetworker — DISMISSED** [review]

File: pkg/operation/botanist/controlplane.go [hunk 2]
```diff
@@ -152,6 +152,8 @@ func (b *Botanist) DeployBackupInfrastructure(ctx context.Context) error {
 		name = common.GenerateBackupInfrastructureName(b.Shoot.SeedNamespace, b.Shoot.Info.Status.UID)
```
Comment (swapnilgm) [inline pkg/operation/botanist/controlplane.go]:
```suggestion
```

File: pkg/operation/botanist/controlplane.go [hunk 3]
```diff
@@ -152,6 +152,8 @@ func (b *Botanist) DeployBackupInfrastructure(ctx context.Context) error {
...
 
 		backupInfrastructure = &gardenv1beta1.BackupInfrastructure{}
+
```
Comment (swapnilgm) [inline pkg/operation/botanist/controlplane.go]:
```suggestion
```

File: pkg/registry/garden/backupinfrastructure/strategy.go [hunk 4]
```diff
@@ -82,6 +87,12 @@ func mustIncreaseGeneration(oldBackupInfrastructure, newBackupInfrastructure *ga
 		return true
 	}
 
+	oldPurpose = oldBackupInfrastructure.ObjectMeta.Annotations[gardencorev1alpha1.GardenPurpose]
```
Comment (swapnilgm) [inline pkg/registry/garden/backupinfrastructure/strategy.go:90]: Also, handle the purpose annotation here: https://github.com/gardener/gardener/blob/master/pkg/registry/garden/shoot/strategy.go#L106.

File: pkg/operation/botanist/controlplane.go [hunk 5]
```diff
@@ -160,8 +162,10 @@ func (b *Botanist) DeployBackupInfrastructure(ctx context.Context) error {
```
Comment (swapnilgm) [inline pkg/operation/botanist/controlplane.go:160]: Since we want to preserve force-deletion annotation; revert the changes at https://github.com/gardener/gardener/pull/1204/files#diff-ac51f64ad7839f6b9db87ac98ae8ced2R166. And add the following line here:

```suggestion
if backupInfrastructure.Annotations == nil {
		backupInfrastructure.Annotations = map[string]interface{}{}
	}
	backupInfrastructure.Annotations[gardencorev1alpha1.GardenPurpose] = shootPurpose
```

**Review by swapnilgm — CHANGES_REQUESTED** [review]
Thank you for the PR.  Please address the suggested changes.

File: pkg/operation/botanist/controlplane.go (hunk 5 above)
Comment (shreyas-s-rao) [reply pkg/operation/botanist/controlplane.go:160]: Agreed. Done.

File: pkg/registry/garden/backupinfrastructure/strategy.go (hunk 4 above)
Comment (shreyas-s-rao) [reply pkg/registry/garden/backupinfrastructure/strategy.go:90]: Done.

File: pkg/controllermanager/controller/backupinfrastructure/backup_infrastructure_control.go (hunk 1 above)
Comment (shreyas-s-rao) [reply pkg/controllermanager/controller/backupinfrastructure/backup_infrastructure_control.go]: Done.

**Review by zanetworker — APPROVED** [review]
